src/transformers/generation/utils.py

```diff
@@ -4756,16 +4838,14 @@ def _speculative_sampling(
     n_matches = ((~is_accepted).cumsum(dim=-1) < 1).sum()  # this is `n` in algorithm 1
 
     # Ensure we don't generate beyond max_len or an EOS token (not in algorithm 1, but needed for correct behavior)
-    if last_assistant_token_is_eos and n_matches == candidate_length:
+    if is_done_candidate and n_matches == candidate_length:
         # Output length is assumed to be `n_matches + 1`. Since we won't generate another token with the target model
         # due to acceptance on EOS we fix `n_matches`
         n_matches -= 1
         valid_tokens = new_candidate_input_ids[:, : n_matches + 1]
     else:
-        n_matches = min(n_matches, max_matches)
-
         # Next token selection: if there is a rejection, adjust the distribution from the main model before sampling.
-        gamma = min(candidate_logits.shape[1], max_matches)
+        gamma = min(candidate_logits.shape[1], n_matches)
```

Comment:
```suggestion
        gamma = candidate_logits.shape[1]
```
`n_matches` is the number of correct guesses, `gamma` is the number of total guesses. `min(candidate_logits.shape[1], n_matches)` would always result in `gamma = n_matches`. In turn, we would always be skipping the `if n_matches < gamma` branch below, which is the key ingredient of speculative decoding :) (see [algorithm 1 on page 3](https://arxiv.org/pdf/2211.17192.pdf))

The original `gamma = min(candidate_logits.shape[1], max_matches)` existed to prevent us from going beyond the EOS token, which the updated `is_done_candidate` line should catch

Reply:
right